server/src/main/java/org/apache/kafka/server/config/AbstractKafkaConfig.java

```diff
@@ -161,6 +169,80 @@ public static Map<String, String> getMap(String propName, String propValue) {
         }
     }
 
+    public static List<Endpoint> listenerListToEndPoints(List<String> listeners, Map<ListenerName, SecurityProtocol> securityProtocolMap) {
+        return listenerListToEndPoints(listeners, securityProtocolMap, true);
+    }
+
+    public static List<Endpoint> listenerListToEndPoints(List<String> listeners, Map<ListenerName, SecurityProtocol> securityProtocolMap, boolean requireDistinctPorts) {
+        List<Endpoint> endPoints;
+        try {
+            endPoints = SocketServerConfigs.listenerListToEndPoints(listeners, securityProtocolMap);
+        } catch (Exception e) {
+            throw new IllegalArgumentException(String.format("Error creating broker listeners from '%s': %s", listeners, e.getMessage()), e);
+        }
+        validate(endPoints, listeners, requireDistinctPorts);
+        return endPoints;
+    }
+
+    private static void validate(List<Endpoint> endPoints, List<String> listeners, boolean requireDistinctPorts) {
+        long distinctListenerNames = endPoints.stream().map(Endpoint::listener).distinct().count();
+        if (distinctListenerNames != endPoints.size()) {
+            throw new IllegalArgumentException("Each listener must have a different name, listeners: " + listeners);
+        }
+
```

Comment:
It can return early if `requireDistinctPorts` is `false`, right? for example:
```java
private static void validate(List<Endpoint> endPoints, List<String> listeners, boolean requireDistinctPorts) {
    long distinct = endPoints.stream().map(Endpoint::listener).distinct().count();
    if (distinct != endPoints.size()) {
        throw new IllegalArgumentException("Each listener must have a different name, listeners: " + listeners);
    }

    if (!requireDistinctPorts) return;

  ...
```